### ISSUE TITLE
[WIP] Adding Support to GL-AX1800 (Flint)

### ISSUE THREAD

**Issue by solidus1983** [solidus1983/openwrt#1]
Adding Support to GL-iNet GL-AX1800 (Flint) to OpenWRT with help from @kirdesde

**Comment by solidus1983** [comment]
Images you made @kirdesde has the reset loop issue.

[serial.txt](https://github.com/solidus1983/openwrt/files/9695398/serial.txt)

**Comment by kirdesde** [comment]
Yeah, please provide a cat from the GPIO with a running TIP image.

But according to the dts they are actually using GPIO 74 as reset.

Can you also flash wie u-boot webinterface the factory.ubi image from my github action? It uses the GPIO 74, maybe the crash only occurs if running from initramfs, but not from flash (just a shot in the dark though)

**Comment by solidus1983** [comment]
i did flash the factory image via uboot but it failed to flash.

**Comment by solidus1983** [comment]
Booted in TIP image now, where am i catting?

**Comment by kirdesde** [comment]
```
cat /sys/kernel/debug/gpio
```

**Comment by solidus1983** [comment]
[Flash.txt](https://github.com/solidus1983/openwrt/files/9695631/Flash.txt)

There is the flash bit and here is the printenv bit
[printenv.txt](https://github.com/solidus1983/openwrt/files/9695640/printenv.txt)

**Comment by kirdesde** [comment]
Looks like they adding a special header (incl. flash script) to their openwrt-FIT image file....

Have you an initramfs image from the TIP repo? Just want to known, if that is booting fine....

**Comment by solidus1983** [comment]
I do have a initramfs build from TIP/GL-iNet Infra repo will upload to mega and link you as that will boot fine without issue.

**Comment by kirdesde** [comment]
No need to upload, I just wanted to know if thats booting fine

**Comment by solidus1983** [comment]
will do bud and thanks, just waiting on a parcel.

**Comment by solidus1983** [comment]
Right here is output with the edit you made.

[output.txt](https://github.com/solidus1983/openwrt/files/9696277/output.txt)

**Comment by kirdesde** [comment]
So at least it's not crashing any more, please post "cat /etc/config/network"

**Comment by kirdesde** [comment]
Ahh okay

**Comment by solidus1983** [comment]
At least we know that gpio 74 causes the reset loop.

**Comment by kirdesde** [comment]
Just for your information:

according to the qsdk kernel dts's the ipq6018-cp03-c1 (same model as the ax1800) is always using gpio pin 75 as reset, not 74.
This is rather strange.....

**Comment by solidus1983** [comment]
ok very strange well the new image has been built lets see how it goes.

**Comment by kirdesde** [comment]
I've no idea any more :-(

**Comment by solidus1983** [comment]
How about we adjust the two changes to 75 instead of 74.

It could also simply be something to do with nss-dp doesn't mrnuke have same issue where there is no traffic going out from LAN.

**Comment by kirdesde** [comment]
We already have tried 75, but feel free to check it.

mrnuke hat another problem regarding external realtek phy, but he has fixed that by updating the ssdk version.

**Comment by solidus1983** [comment]
ah right, got ya. Will compiling now, it seems strange that Wireless works but not ethernet, it should be the other way round.

**Comment by solidus1983** [comment]
Still nothing with that adjustment, i wonder if we are missing firmware for the phy.

**Comment by kirdesde** [comment]
For ethernet only qca-nss-dp and qca-ssdk is necessary, ssdk also contains all the network drivers for the phy (qca8075)

Wifi on the other hand uses ath11k and the wifi interfaces are connected through the br-lan bridge.
So the lan-br bridge is working as wifi dhcp works.

Let me think about it ....

**Comment by solidus1983** [comment]
dude i have a ping.

**Comment by solidus1983** [comment]
yes via ethernet

**Comment by kirdesde** [comment]
WTH, and the other eth ports are also working?
And what about eth0 as wan port, is that working?

**Comment by solidus1983** [comment]
only LAN4 is working.

**Comment by kirdesde** [comment]
you can try to change this line in the dts from:

```
switch_lan_bmp = <0x3c>; /* lan port bitmap */
```
to
```
switch_lan_bmp = <0x1e>; /* lan port bitmap */
```
This changes the port bitmap of the switch.

**Comment by solidus1983** [comment]
ok will do give me a second need to clear the build server for a clean build as i made off cuff edit.

**Comment by kirdesde** [comment]
You could also try <&tlmm 75 GPIO_ACTIVE_LOW>; as mdio reset....

**Comment by solidus1983** [comment]
Ok will make the change now

**Comment by solidus1983** [comment]
Made the change built and same thing happened however changing eth4 to eth0 and eth0 to eth4 gives me a ping again on LAN4 port.

**Comment by kirdesde** [comment]
> Ok will make the change now
> 
> I did notice that TIP has
> 
> `switch_lan_bmp = <0x0c>; /* lan port bitmap */` under ess-switch

Yes in the dtsi, but thats get's overritten by the dts content.

**Comment by kirdesde** [comment]
Can yo post /etc/config/network from TIP running image?

**Comment by solidus1983** [comment]
root@GL-AX1800:~# cat /etc/config/network

config interface 'loopback'
        option device 'lo'
        option proto 'static'
        option ipaddr '127.0.0.1'
        option netmask '255.0.0.0'

config globals 'globals'
        option ula_prefix 'fd6c:1471:e1bf::/48'

config device
        option name 'br-lan'
        option type 'bridge'
        list ports 'eth1'
        list ports 'eth2'
        list ports 'eth3'
        list ports 'eth4'
        option multicast_to_unicast '0'

config device
        option name 'eth1'
        option macaddr 'd0:0d:fe:ed:01:33'

config device
        option name 'eth2'
        option macaddr 'd0:0d:fe:ed:01:33'

config device
        option name 'eth3'
        option macaddr 'd0:0d:fe:ed:01:33'

config device
        option name 'eth4'
        option macaddr 'd0:0d:fe:ed:01:33'

config interface 'lan'
        option device 'br-lan'
        option proto 'static'
        option ipaddr '192.168.8.1'
        option netmask '255.255.255.0'
        option ip6assign '60'
        option isolate '0'

config device
        option name 'eth0'
        option macaddr 'd0:0d:fe:ed:01:32'

config interface 'wan'
        option device 'eth0'
        option proto 'dhcp'
        option ipv6 '0'

config interface 'wan6'
        option device 'eth0'
        option proto 'dhcpv6'
        option disabled '1'

config switch
        option name 'switch0'
        option reset '0'
        option enable_vlan '0'

config interface 'guest'
        option force_link '1'
        option type 'bridge'
        option proto 'static'
        option ipaddr '192.168.9.1'
        option netmask '255.255.255.0'
        option ip6assign '60'
        option multicast_querier '0'
        option igmp_snooping '0'
        option isolate '0'
        option bridge_empty '1'

config interface 'wwan'
        option proto 'dhcp'

config rule 'policy_bypass_vpn'
        option mark '0x60000/0x60000'
        option lookup '53'
        option priority '53'

config rule 'policy_via_vpn'
        option mark '0x80000/0x80000'
        option lookup '52'
        option priority '52'

config rule 'policy_dns'
        option mark '0x100000/0x100000'
        option lookup '51'

**Comment by solidus1983** [comment]
thats on GL-iNet's v4 firmware which uses TIP branch.

**Comment by solidus1983** [comment]
On it right now.

**Comment by kirdesde** [comment]
If that's still not work try mdio reset GPIO 74 HIGH as last resort....
I'm away from the PC for a couple of ours now.

**Comment by solidus1983** [comment]
Yeah no problem will try the suggestions you have made, Thanks i will be away from computer soon as well.

**Comment by solidus1983** [comment]
Right with that change on 02_network LAN4 will ping after ip change no issue else LAN1-3 are still not reachable via ping.

Will check WAN and report back

**Comment by solidus1983** [comment]
Update: Looking at GL-iNet's TIP build their 02_network shows

glinet,ax1800)
                ucidef_set_interface_lan "eth1 eth2 eth3 eth4"
                ucidef_set_interface_wan "eth0"
                ;;

**Comment by kirdesde** [comment]
> Update: Looking at GL-iNet's TIP build their 02_network shows
> 
 ```
> glinet,ax1800)
>                 ucidef_set_interface_lan "eth1 eth2 eth3 eth4"
>                 ucidef_set_interface_wan "eth0"
>                 ;;
>
```
 This is just the old (deprecated) syntax

> But also has
> 
 ```
>     local board="$1"
>     ucidef_add_switch "switch0"
>     ucidef_add_switch_attr "switch0" "enable" "false"
>     ucidef_add_switch_attr "switch0" "reset" "false"
 ```

->  This is the old switchdev syntax, we have never used this for ipq807x (same for ipq60xx) and it's even deprecated as well in OpenWrt.

> at the top
> 
> New update: changed both 75's to 74 [bd3349a](https://github.com/solidus1983/openwrt/commit/bd3349ade3e571caa3332fab619f38becd9b47aa)
> 
> no reboot

**Comment by solidus1983** [comment]
Thanks for reply, tried everything (check my commits) still the same result of LAN4 is the only port that works right now.

**Comment by kirdesde** [comment]
Hi, I'm playing with a Netgaer WAX610 and seeing the same crash during phy init like you.
This was not the case with an older build.

Something has changed during the last commit in robimarkos repo.

**Comment by kirdesde** [comment]
I pushed a fix with a phy-reset delay and I've also changed the 02_network.

Please try that /(eave the rest at 74 HIGH)

**Comment by solidus1983** [comment]
right i have made the commit changes manually as our repo's have become unsync'd so it wanted me to discard 12 commits.
Will compile and report back to you.

**Comment by solidus1983** [comment]
Right with 74 set to high its still reboot looping.

**Comment by solidus1983** [comment]
[rebootloop.txt](https://github.com/solidus1983/openwrt/files/9701030/rebootloop.txt)

There is the output, it seems to kernel panic then reboot.

**Comment by solidus1983** [comment]
@kirdesde Just seen your commit so i have merged it, will compile right now.

**Comment by solidus1983** [comment]
Still reboot looped from kernel panic, however setting it back to LOW on reset gpio section boots, like are we missing something.

**Comment by kirdesde** [comment]
This is robimarkos recommendation from the other issue:

Hm, I would then try poking around GPIO-s manually with gpiod and using mdio-tools to see when PHY-s pop up.
I dont really have a better idea, but QCA807x for sure is reset when its reset line is pulled low.

So you would need mdio-tools in the initramfs image and play with the reset gpio's.

You first need to export them and then can change the value:

https://openwrt.org/docs/techref/hardware/port.gpio

**Comment by solidus1983** [comment]
So rebuild with mdio-tools added in .config then use that guide to find gpio's got it.

**Comment by solidus1983** [comment]
`root@OpenWrt:/# cat /sys/class/gpio/gpiochip*/base | head -n1
432
root@OpenWrt:/# echo "506" > /sys/class/gpio/export
ash: write error: Resource busy
`
Humm

**Comment by solidus1983** [comment]
I am right that its Base vaule 432 + 74 (GPIO) or am i doing that wrong.

**Comment by kirdesde** [comment]
That should be right. But please note, GPIO is already hooked up to the phy reset as per DTS definition.

So you would first need to remove that from the DTS

**Comment by solidus1983** [comment]
ok got ya, so both lines with 74 or just the pull up line.

**Comment by solidus1983** [comment]
Might have found why its doing what its doing

`qcom,ipq6018-ap-cp02-c1 |\
	qcom,ipq6018-ap-cp03-c1 |\
	qcom,ipq6018-db-cp02)
		ucidef_set_interfaces_lan_wan "eth1" "eth0"
		;;
`

**Comment by solidus1983** [comment]
Ok so i have confirmed that 74 is the phy as with the gpio-sysfs if i type

`./gpiosys 74 out 1` it results with `[  712.070992] nss-dp 3a001400.dp eth2: PHY Link up speed: 1000` doing `./gpiosys 74 out 0` it results with '[  755.750793] nss-dp 3a001400.dp eth2: PHY Link is down` .

So even more clueless now.

**Comment by robimarko** [comment]
Not really, BTW, you cannot change direction to input and look for value you set as output.
As soon as you switch it to input, it will be HighZ and pull-up or pull-down will do its thing, in this case it should be high as there should be a pull-up so that PHY doesn't randomly reset

**Comment by solidus1983** [comment]
Thanks for your reply robi, guess it back to drawing board, then still can't understand why booting with pin 74 set to HIGH causes a kernel panic reset loop, whilst setting pin 74 to LOW it boots fine.

Thanks for your input bud.

**Comment by robimarko** [comment]
Can you post the whole DTS?

**Comment by solidus1983** [comment]
DTS: https://github.com/solidus1983/openwrt/blob/eap610-outdoor/target/linux/ipq60xx/files/arch/arm64/boot/dts/qcom/ipq6018-gl-ax1800.dts

DTSI: https://github.com/solidus1983/openwrt/blob/eap610-outdoor/target/linux/ipq60xx/files/arch/arm64/boot/dts/qcom/ipq6018-gl-ax1800.dtsi

**Comment by robimarko** [comment]
Well, on a glimpse I dont see anything broken except for the reset polarity

**Comment by solidus1983** [comment]
Ok, which line and what to change. I can then compile it and then test.

**Comment by robimarko** [comment]
https://github.com/solidus1983/openwrt/blob/eap610-outdoor/target/linux/ipq60xx/files/arch/arm64/boot/dts/qcom/ipq6018-gl-ax1800.dtsi#L139

To `GPIO_ACTIVE_HIGH` and remove that ultra-long delay, QCA807x will come up instantly

**Comment by robimarko** [comment]
Forgot to edit after copy/pasting change it to `GPIO_ACTIVE_LOW` and remove the delay.

**Comment by solidus1983** [comment]
No problem.

**Comment by solidus1983** [comment]
@kirdesde where did you get

`compatible = "ethernet-phy-id004d.d101";`

from?

**Comment by robimarko** [comment]
Its for the QCA8081, I dont think you have of those.
Its assembled per the ethernet PHY bindings

**Comment by solidus1983** [comment]
@robimarko  it was just a general question as GL-AX1800 is a QCA8075.

**Comment by robimarko** [comment]
Its defined here:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/devicetree/bindings/net/ethernet-phy.yaml?h=v5.15.72#n30

You need it if you put the reset under the PHY

**Comment by solidus1983** [comment]
Ok thanks for the doc, will have a read, just something is nagging me with the -c22 bit.

**Comment by robimarko** [comment]
C22 is correct as QCA8075 is a Clause-22 compliant PHY

**Comment by solidus1983** [comment]
Thanks for that, still hitting the LAN4 port being the only one fully working and LAN1-3 as well as WAN only be able to detect connect and disconnects of the cable with no ability to give or receive DHCP ip's

**Comment by kirdesde** [comment]
Are you seeing this by mdio-tools?

**Comment by solidus1983** [comment]
no dmesg and cat'ing sys/class/net/<eth>/carrier

**Comment by solidus1983** [comment]
Right mdio shows LAN1 up which is where my cable is connected to nss-dp is still resetting eth1.

**Comment by robimarko** [comment]
You are gonna have to explain the "resetting" part

**Comment by robimarko** [comment]
Yeah, it's not a PHY issue but without a crystal ball its impossible to guess if SSDK or NSS-DP are broken

**Comment by solidus1983** [comment]
I disabled TSO and GSO it still doing it.

**Comment by kirdesde** [comment]
If I'm not wrong, I had this DP enable/disable issue on IPQ807x as well, but at the moment I can't remember that the reason was :-(

**Comment by kirdesde** [comment]
@robimarko: could this be a power related issue?

I've just removed the upper opp steps in the ax1800.dtsi

```
&cpu_opp_table {
	/delete-node/ opp-1440000000;
	/delete-node/ opp-1608000000;
	/delete-node/ opp-1800000000;	
	};
```
But I didn't changed the core voltage steps, so the IPQ6000 SOC runs with the same opp steps as the IPQ6018 (but only up to 1.3 Ghz) as defined in the upstream ipq6018.dts

In the dtsi from the TIP the using a fixed core voltage of 1.1 volt for all opp steps.

**Comment by robimarko** [comment]
Its hard to tell, but it could be, any predictable behaviour could be voltage or clock issue.

But yeah, OPP-s that are present are meant only for 6010/18 SoC-s, there are also steps missing in them

**Comment by solidus1983** [comment]
Right @kirdesde even after deleting the openwrt folder and starting from scratch and also parse ```make defconfig```

```
root@OpenWrt:/# cat /sys/kernel/debug/regulator/regulator_summary
 regulator                      use open bypass  opmode voltage current     min     max
---------------------------------------------------------------------------------------
 regulator-dummy                  1    2      0 unknown     0mV     0mA     0mV     0mV
    59000.qusb-vdda-phy-dpdm      0                                 0mA     0mV     0mV
    59000.qusb-vdda-pll           0                                 0mA     0mV     0mV
```

**Comment by robimarko** [comment]
`make kernel_menuconfig` is your friend

**Comment by solidus1983** [comment]
in there already but using nconfig instead.

**Comment by solidus1983** [comment]
@kirdesde how does one make this flashable via uboot btw? As for some reason i can't use uboot GUI to flash it directly to the router.

**Comment by kirdesde** [comment]
Have a look in the header of the openwrt-ax1800-4.1.0-0914-beta1.img

```
888(K8;c!Flashing factory image^LimagesscriptscriptDcho '######################################## Flashing ubi: St
arted'
setenv stdout nulldev
smem
imxtract $imgaddr ubi || setenv stdout serial && echo "invalid image" && exit 1
setenv stdout serial
nand erase.part fs
nand write $fileaddr fs $filesize
echo '######################################## Flashing ubi: Done'
exit 0 script%ARM*nonehash@1EIlF6crc32ubiubi8BIc!]ʕ
```
They have also defined some flash scripts in uboot, you can see them via "printenv"

But I'd rather use our "standard" way to flash, boot initramfs, scp the openwrt factory.ubi and do a ```ubiformat /dev/mtdX -y -f /path/to/factory.ubi```
There /dev/mtdX is the rootfs partition.
